synapse/storage/background_updates.py

```diff
@@ -758,6 +758,11 @@ def create_index_psql(conn: Connection) -> None:
                 logger.debug("[SQL] %s", sql)
                 c.execute(sql)
 
+                # override the global statement timeout to avoid accidentally squashing
+                # a long-running index creation process
+                timeout_sql = "SET LOCAL statement_timeout = 0"
+                c.execute(timeout_sql)
```

Comment:
We need to set this back to the default, as we're not in a transaction when creating indices I think?

Reply:
Is the autocommit being on the reason that this isn't in a transaction? The statement timeout only applies for a given transaction - if we are *not* in a transaction then I wonder if it will have any effect. I will do some digging in the docs but they were pretty scarce on information when I looked into this the first time. Do we have any Postgres/psychopg wizards around who might have some arcane insight into this?

Reply:
> Is the autocommit being on the reason that this isn't in a transaction?

Yes



> The statement timeout only applies for a given transaction - if we are _not_ in a transaction then I wonder if it will have any effect.

https://www.postgresql.org/docs/current/sql-set.html says that `LOCAL` doesn't work outside a transaction, and so we'd need to use `SESSION`